lib/spack/llnl/util/filesystem.py

```diff
@@ -1746,56 +1753,38 @@ def _log_file_access_issue(e: OSError, path: str) -> None:
     tty.debug(f"find must skip {path}: {errno_name} {e}")
 
 
-@system_path_filter(arg_slice=slice(1))
-def find_max_depth(root, globs, max_depth: Optional[int] = None):
-    """Given a set of non-recursive glob file patterns, finds all
-    files matching those patterns up to a maximum specified depth.
-
-    If a directory has a name which matches an input pattern, it will
-    not be included in the results.
-
-    If ``max_depth`` is specified, does not search below that depth.
-
-    If ``globs`` is a list, files matching earlier entries are placed
-    in the return value before files matching later entries.
-    """
-    try:
-        stat_root = os.stat(root)
-    except OSError:
-        return []
-
-    if max_depth is None:
-        max_depth = sys.maxsize
+def _dir_id(s: os.stat_result) -> Tuple[int, int]:
+    # Note: on windows, st_ino is the file index and st_dev is the volume serial number. See
+    # https://github.com/python/cpython/blob/3.9/Python/fileutils.c
+    return (s.st_ino, s.st_dev)
+
+
+def _find_max_depth(roots: List[str], globs: List[str], max_depth: int = sys.maxsize) -> List[str]:
+    """See ``find`` for the public API."""
+    # Apply normcase to file patterns and filenames to respect case insensitive filesystems
+    to_regex = lambda x: fnmatch.translate(os.path.normcase(x))
+    # Create a single regex where pattern i is matched by group p{i}
+    regex = re.compile("|".join(rf"(?P<p{i}>{to_regex(x)})" for i, x in enumerate(globs)))
```

Comment:
I mentioned this should be factored into a library function. I still feel that way. If you would prefer to do that elsewhere, can you restore the original check (with the added break statement for avoiding duplicates)?